### PR TITLE
Serialization improvements

### PR DESCRIPTION
**Pull request by vpodlesnyak**
This pull request uses serializers API to simplify encoding / decoding and provides more type safe way to operate with Timestamps

### CODE REVIEW

File: firebase-common/src/androidMain/kotlin/dev/gitlive/firebase/_decoders.kt [hunk 1]
```diff
@@ -9,15 +9,15 @@ import kotlinx.serialization.descriptors.PolymorphicKind
 import kotlinx.serialization.descriptors.SerialDescriptor
 import kotlinx.serialization.descriptors.StructureKind
 
-actual fun FirebaseDecoder.structureDecoder(descriptor: SerialDescriptor, decodeDouble: (value: Any?) -> Double?): CompositeDecoder = when(descriptor.kind) {
```
Comment (vpodlesnyak) [inline firebase-common/src/androidMain/kotlin/dev/gitlive/firebase/_decoders.kt:12]: special value handling is replaced by using a custom serializer

File: firebase-common/src/commonMain/kotlin/dev/gitlive/firebase/serializers.kt [hunk 2]
```diff
@@ -104,3 +104,30 @@ class FirebaseListSerializer : KSerializer<Iterable<Any?>> {
     }
 }
 
+/**
+ * A special case of serializer for values natively supported by Firebase and
+ * don't require an additional encoding/decoding.
+ */
+abstract class SpecialValueSerializer<T>(
```
Comment (vpodlesnyak) [inline firebase-common/src/commonMain/kotlin/dev/gitlive/firebase/serializers.kt:111]: in this PR it's used for Realtime Database `ServerValue` and Firestore `FieldValue` and `Timestamp`. _Firestore also supports storing `GeoPoints` and `DocumentReferences` natively, but it's out of scope of this PR_

File: firebase-database/src/androidMain/kotlin/dev/gitlive/firebase/database/ServerValue.kt [hunk 3]
```diff
@@ -0,0 +1,21 @@
+package dev.gitlive.firebase.database
+
+import kotlinx.serialization.Serializable
+
+private typealias NativeServerValue = com.google.firebase.database.ServerValue
+
+/** Represents a Firebase ServerValue. */
+@Serializable(with = ServerValueSerializer::class)
+actual class ServerValue internal actual constructor(
```
Comment (vpodlesnyak) [inline firebase-database/src/androidMain/kotlin/dev/gitlive/firebase/database/ServerValue.kt:9]: using a wrapper class allows specifying a custom serializer and also ensures correct equality checks in JS

File: firebase-firestore/src/androidMain/kotlin/dev/gitlive/firebase/firestore/firestore.kt [hunk 4]
```diff
@@ -118,27 +118,15 @@ actual class WriteBatch(val android: com.google.firebase.firestore.WriteBatch) {
 
     @JvmName("updateFields")
     actual fun update(documentRef: DocumentReference, vararg fieldsAndValues: Pair<String, Any?>) =
-        android.takeUnless { fieldsAndValues.isEmpty() }
-            ?.update(
-                documentRef.android,
-                fieldsAndValues[0].first,
```
Comment (vpodlesnyak) [inline firebase-firestore/src/androidMain/kotlin/dev/gitlive/firebase/firestore/firestore.kt:124]: a bugfix here. the first value was never encoded

File: firebase-firestore/src/commonMain/kotlin/dev/gitlive/firebase/firestore/Timestamp.kt [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    fromNativeValue = { value ->
+        when (value) {
+            FieldValue.serverTimestamp.nativeValue -> Timestamp.ServerTimestamp
+            else -> throw SerializationException("Cannot deserialize $value")
+        }
+    }
+)
+
+/** A serializer for a Double field which is stored as a Timestamp. */
+object DoubleAsTimestampSerializer : SpecialValueSerializer<Double>(
```
Comment (vpodlesnyak) [inline firebase-firestore/src/commonMain/kotlin/dev/gitlive/firebase/firestore/Timestamp.kt:82]: A serializer which allows storing a timestamp in a double field. potentially useful for a legacy behaviour

File: firebase-firestore/src/iosMain/kotlin/dev/gitlive/firebase/firestore/firestore.kt [hunk 6]
```diff
@@ -112,10 +99,16 @@ actual class WriteBatch(val ios: FIRWriteBatch) {
         ios.updateData(encode(strategy, data, encodeDefaults) as Map<Any?, *>, documentRef.ios).let { this }
 
     actual fun update(documentRef: DocumentReference, vararg fieldsAndValues: Pair<String, Any?>) =
-            ios.updateData(fieldsAndValues.associate { it }, documentRef.ios).let { this }
+        ios.updateData(
+            fieldsAndValues.associate { (field, value) -> field to encode(value, true) },
```
Comment (vpodlesnyak) [inline firebase-firestore/src/iosMain/kotlin/dev/gitlive/firebase/firestore/firestore.kt:103]: a bugfix here. it seems like iOS was not encoding values

File: firebase-database/src/commonTest/kotlin/dev.gitlive.firebase.database/database.kt [hunk 7]
```diff
@@ -65,6 +65,42 @@ class FirebaseDatabaseTest {
         assertEquals(3, firebaseDatabaseChildCount)
     }
 
+    @Test
```
Comment (vpodlesnyak) [inline firebase-database/src/commonTest/kotlin/dev.gitlive.firebase.database/database.kt:68]: _I'd appreciate some help on running JS tests for `firebase-database` module locally. there is a warning on `jsLegacyTest` run on current `master`_
```
ReferenceError: $module$firebase_compat_app is not defined
    at ..../firebase-database/src/jsMain/kotlin/dev/gitlive/firebase/database/database.kt:18:1
    at Object.<anonymous> (..../build/js/packages/firebase-kotlin-sdk-firebase-database-js-legacy-test/kotlin/firebase-kotlin-sdk-firebase-database-js-legacy-test.js:1093:2)
```
_and it seems like the tests aren't executed at all - no reports, no DB updates_

**Comment by Reedyuk** [comment]
can you rebase and then we can re-review it

**Comment by vpodlesnyak** [comment]
@Reedyuk done

**Comment by Reedyuk** [comment]
Looks good, lets wait for the ci to pass and @nbransby can review too

**Pull request merged**